### PR TITLE
feature (db): Add support for multi-column PKs

### PR DESCRIPTION
**Pull request by jimlambrt**
Add support for the db layer operations like lookup, delete, update,
etc to support resources with mulit-column primary keys.  It also
adds support for PKs which are not either public_id or private_id.

These new capabilities are needed for things in #1634 ( FYI: @talanknight @vancluever)

### CODE REVIEW

File: internal/db/schema/migrations/oss/postgres/19/01_db.up.sql [hunk 1]
```diff
@@ -0,0 +1,56 @@
+begin;
```
Comment (jefferai) [inline internal/db/schema/migrations/oss/postgres/19/01_db.up.sql:1]: Should this be in a migration for everyone? Or should it be part of the tests utilizing it?

**Review by jefferai — COMMENTED** [review]
Looks pretty good. There's a failing test and I'm casting some side-eye at that migration but outside of that the approach looks solid.

File: internal/db/read_writer_test.go [hunk 2]
```diff
@@ -2013,16 +2061,34 @@ func TestDb_LookupById(t *testing.T) {
...
+			name:       "simple-non-public-non-private-id",
+			underlying: db,
+			args: args{
+				resource: accessory,
+			},
+			wantErr: false,
+			want:    accessory,
+		},
+		{
+			name:       "compond",
```
Comment (mgaffney) [inline internal/db/read_writer_test.go:2079]: Can we add a test to verify an error is returned when one or more values of a compound key are the zero value?

**Review by mgaffney — CHANGES_REQUESTED** [review]
Nice work! Just one request for an additional test.

File: internal/db/schema/migrations/oss/postgres/19/01_db.up.sql (hunk 1 above)
Comment (jimlambrt) [reply internal/db/schema/migrations/oss/postgres/19/01_db.up.sql:1]: I agree, that is shouldn't be.  With that said, this PR is aligned with the current pattern used by internal/db for test schema objects.  I've opened an issue to refactor all of them: https://hashicorp.atlassian.net/browse/ICU-2841 in a separate PR.

File: internal/db/read_writer_test.go (hunk 2 above)
Comment (jimlambrt) [reply internal/db/read_writer_test.go:2079]: ty!

**Comment by jimlambrt** [comment]
> Looks pretty good. There's a failing test and I'm casting some side-eye at that migration but outside of that the approach looks solid.

Working on the failed test.  happy to say, the failure shed light on some areas for improvement in that set of unit tests, which led to finding/fixing a new issue.

**Review by jefferai — APPROVED** [review]

**Review by mgaffney — APPROVED** [review]
